std/experimental/checkedint.d

```diff
@@ -0,0 +1,1626 @@
+        }
+        else static if (hasMember!(Hook1, "hookOpBinary"))
+        {
+            // Delegate to rhs
+            return mixin("this.payload"~op~"rhs");
+        }
+        else static if (hasMember!(Hook, "onOverflow") &&
+            !hasMember!(Hook1, "onOverflow"))
+        {
+            // Delegate to lhs
+            return mixin("this"~op~"rhs.payload");
+        }
+        else static if (hasMember!(Hook1, "onOverflow") &&
+            !hasMember!(Hook, "onOverflow"))
+        {
+            // Delegate to rhs
+            return mixin("this.payload"~op~"rhs");
+        }
+        else
+        {
+            static assert(0, "Conflict between lhs and rhs hooks,"
+                " use .representation on one side to disambiguate.");
+        }
+    }
+
+    // opBinaryRight
+    /**
+    */
+    auto opBinaryRight(string op, Lhs)(const Lhs lhs)
+    if (isIntegral!Lhs || isFloatingPoint!Lhs || is(Lhs == bool))
+    {
+        static if (hasMember!(Hook, "hookOpBinaryRight"))
+        {
+            auto r = hook.hookOpBinaryRight!op(lhs, payload);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else static if (hasMember!(Hook, "hookOpBinary"))
+        {
+            auto r = hook.hookOpBinary!op(lhs, payload);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else static if (is(Lhs == bool))
+        {
+            return mixin("ubyte(lhs)"~op~"this");
+        }
+        else static if (isFloatingPoint!Lhs)
+        {
+            return mixin("lhs"~op~"payload");
+        }
+        else static if (hasMember!(Hook, "onOverflow"))
+        {
+            bool overflow;
+            auto r = opChecked!op(lhs, T(payload), overflow);
+            if (overflow) r = hook.onOverflow!op(42);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else
+        {
+            // Default is built-in behavior
+            auto r = mixin("lhs"~op~"T(payload)");
+            return Checked!(typeof(r), Hook)(r);
+        }
+    }
+
+    // opOpAssign
+    /**
+    */
+    ref Checked opOpAssign(string op, Rhs)(const Rhs rhs)
+    if (isIntegral!Rhs || isFloatingPoint!Rhs || is(Rhs == bool))
+    {
+        static assert(is(typeof(mixin("payload"~op~"=rhs")) == T));
+
+        static if (hasMember!(Hook, "hookOpOpAssign"))
+        {
+            hook.hookOpOpAssign!op(payload, rhs);
+        }
+        else
+        {
+            alias R = typeof(payload + rhs);
+            auto r = mixin("this"~op~"rhs").payload;
+
+            static if (valueConvertible!(R, T) ||
+                !hasMember!(Hook, "onBadOpOpAssign") ||
+                op.among(">>", ">>>"))
+            {
+                // No need to check these
+                payload = cast(T) r;
+            }
+            else
+            {
+                static if (isUnsigned!T && !isUnsigned!R)
+                    // Example: ushort += int
+                    const bad = r < 0 || r > max.payload;
```

Comment:
This is equivalent to `const bad = unsigned(r) > max.payload;` I'm not sure if optimizers catch it or not.